File: solutions/javascript/dfs/allPathsFromSourceToTarget.js
const START_NODE = 0;

const getPathsFromNodeToTarget = (graph, curNode, targetNode) => {
    // Base cases
    if (curNode === targetNode) return [[curNode]];

    // Process node
    const allPathsFromNodeToTarget = [];

    // Recurse on neighbors
    const neighbors = graph[curNode];
    for (const neighbor of neighbors) {
        const pathsFromNeighborToTarget = getPathsFromNodeToTarget(
            graph,
            neighbor,
            targetNode,
        );

        for (const path of pathsFromNeighborToTarget) {
            const pathFromCurToTarget = [curNode, ...path];

            allPathsFromNodeToTarget.push(pathFromCurToTarget);
        }
    }

    return allPathsFromNodeToTarget;
};

const allPathsSourceTarget = (graph) => {
    const targetNode = graph.length - 1;

    return getPathsFromNodeToTarget(graph, START_NODE, targetNode);
};

File: solutions/python/dfs/all_paths_from_source_to_target.py

START_NODE = 0

class Solution:
    def get_paths_from_node_to_target(self, graph, cur_node, target_node):
        # Base cases
        if cur_node == target_node:
            return [[cur_node]]

        # Process node
        all_paths_from_node_to_target = []

        # Recurse on neighbors
        neighbors = graph[cur_node]
        for neighbor in neighbors:
            paths_from_neighbor_to_target = self.get_paths_from_node_to_target(
                graph, neighbor, target_node)

            for path in paths_from_neighbor_to_target:
                path_from_cur_to_target = [cur_node, *path]

                all_paths_from_node_to_target.append(path_from_cur_to_target)

        return all_paths_from_node_to_target

    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        target_node = len(graph) - 1

        return self.get_paths_from_node_to_target(graph, START_NODE, target_node)